library/init/meta/coinductive_predicates.lean

```diff
@@ -0,0 +1,326 @@
+/-
+Copyright (c) 2017 Johannes Hölzl. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Johannes Hölzl (CMU)
+-/
+prelude
+import init.meta.expr init.meta.tactic init.meta.constructor_tactic
+
+namespace name
+
+def last_string : name → string
+| anonymous        := "[anonymous]"
+| (mk_string s _)  := s
+| (mk_numeral _ n) := last_string n
+
+end name
+
+namespace expr
+open expr
+
+meta def replace_with (e : expr) (s : expr) (s' : expr) : expr :=
+e.replace $ λc d, if c = s then some s' else none
+
```

Comment:
`replace_with` will produce incorrect results if `s'` contains dangling de-Bruijn variables.
This is not an issue in your code since this case never occurs.
However, if the procedure `replace_with` is available, other users may invoke it with terms containing dangling deBruijn variables. We can mark it private, or use `lift_vars`.
```lean
meta def replace_with (e : expr) (s : expr) (s' : expr) : expr :=
e.replace $ λc d, if c = s then some (s'.lift_vars 0 d) else none
```